feat: added `tests/typescript` to test supported TS examples

btw17 opened the pull request:
<!--
  - Thanks for submitting code to Blockly!  Please fill out the following as part of
  - your pull request so we can review your code more easily.
  -->

## The basics

<!-- TODO: Verify the following, checking each box with an 'x' between the brackets: [x] -->

- [x] I branched from develop
- [x] My pull request is against develop
- [x] My code follows the [style guide](https://developers.google.com/blockly/guides/modify/web/style-guide)
- [x] I ran `npm run format` and `npm run lint`

## The details
### Resolves

This adds support for testing theoretical Blockly use cases, as mentioned in https://github.com/google/blockly/pull/6639#pullrequestreview-1212386961

### Proposed Changes

Updated `npm run test` to include a new suite of tests connected to `tests/typescript`:
- The test passes if everything there compiles.
- The test fails if the files fail to compile.
  - The error output from the `tsc` command displays in the console output.

#### Behavior Before Change

One less test suite to run.

#### Behavior After Change

One more test suite to run.

### Reason for Changes

To enable testing fake custom fields. This allows us to see how core updates impact use cases we'd like to support.

### Test Coverage

N/A

### Documentation

I don't believe so.

### Additional Information

I have one example test case that's related to a discussion in #6639 about allowing different user input from

cpcallen reviewed (changes requested):
Yay for more tests!

btw17 commented on .eslintignore:
```diff
@@ -11,6 +11,7 @@
 /tests/screenshot/*
 /tests/test_runner.js
 /tests/workspace_svg/*
+/tests/typescript/*
```
I'm not sure how straightforward it will be, though I'll look further into it this upcoming Friday. And get back to you when I've looked into it a bit more.

The configuration in `.eslintrc.json` points to the root `./tsconfig.json` via  `parserOptions` which I believe was the main issue, though there may be around that. These tests need their own tsconfig so they aren't built and packaged with the rest of the code.

btw17 replied:
Updated! The tsconfig is defined in an array, which I hadn't noticed at my quick glance earlier, so I just had to add the second override.

Of note, I updated `.eslintrc.json` to `.eslintrc.js` since we likely want the same settings for the overrides, and it's easier to keep them in sync via JS than in JSON.

BeksOmega commented on tests/typescript/README.md:
```diff
@@ -0,0 +1,4 @@
+# TypeScript Tests
+
+Tests pass if all files here compile. Add Blockly use case examples we'd like to support, such as
+subclasses which extend from Blockly classes.
```
There actually is a style guide! https://google.github.io/styleguide/docguide/style.html

btw17 commented on scripts/gulpfiles/test_tasks.js:
```diff
@@ -342,6 +342,23 @@ function advancedCompileInBrowser() {
+  const id = 'supported_typescript_examples';
+  const command = 'tsc -p ./tests/typescript/tsconfig.json';
+  return runTestTask(id, async() => {
+    return new Promise((resolve, reject) => {
+      // NOTE: execSync throws with `err` as the message, though
+      // `stdout` is needed to display the tsc error output.
+      exec(command, (err, stdout) => {
+        const tscFailed = !!err;
+        const tscLog = stdout;
+        if (tscFailed) reject(tscLog);
+        resolve();
+      });
```
Thanks for double checking this! I do not need to suppress the main error, and on further review of the output I can probably just use `runTestCommand` as is. Given the three cases: `spawn`, `execSync` (via `runTestCommand`), and my current code:

- `spawn` does error out with the `stdout` text included, though it immediately stops the test process without proceeding to further steps. (It never reaches `reportTestResult`)
- `runTestCommand` only displays the `err` output at the bottom in the summary (basically just stating that the `tsc` command failed), though it *does* actually include the `stdout` text above in the log at the time the test is actually ran.
- My code includes the `stdout` text (which is the important stuff to debug a failure) in the bottom in the summary, though it only displays `undefined` when the test actually ran.

Having the `stdout` included above in the log is probably sufficient.

btw17 commented on scripts/gulpfiles/test_tasks.js:
```diff
@@ -352,6 +369,7 @@ const tasks = [
   mocha,
   generators,
   node,
+  supportedTypescriptExamples,
```
That makes sense. I'll name it `typeDefinitions` since we are testing that the Blockly type definitions work with the supported TypeScript examples.

btw17 commented on scripts/gulpfiles/test_tasks.js:
```diff
@@ -342,6 +342,23 @@ function advancedCompileInBrowser() {
   return runTestTask('advanced_compile_in_browser', runCompileCheckInBrowser);
 }
 
+function supportedTypescriptExamples() {
```
Added!

btw17 commented on tests/typescript/src/field/different_user_input.ts:
```diff
@@ -0,0 +1,75 @@
+            const rightCell: Cell = {cellId: `${cell.cellId}-right`};
+            return [leftCell, rightCell]
+        }).reduce((allCells, cellDuo) => allCells.concat(cellDuo), []);
+
```
Sounds good. I'll remove the blank line between defining the cells and setting the value.

btw17 commented on tests/typescript/src/field/different_user_input.ts:
```diff
@@ -0,0 +1,75 @@
+/**
```
Added!

btw17 commented on tests/typescript/src/field/different_user_input.ts:
```diff
@@ -0,0 +1,75 @@
+    cells: Cell[];
+}
+
+type FieldMitosisValidator = FieldValidator<CellGroup>;
```
I'm no biology expert, though I thought this would be a fun example. Basically just cells multiplying over time by splitting. Is this example fine?

cpcallen replied:
Yes, absolutely.  I was just lamenting that I don't really remember any biology, seeing as I last did any in grade ten, which was quite a lot of decades ago now…

btw17 commented on tests/typescript/tsconfig.json:
```diff
@@ -0,0 +1,22 @@
+    "allowJs": true,
+    "outDir": "dist",
+    "baseUrl": ".",
+    "paths": {
+      "blockly-test/*": ["../../dist/*"],
+    },
```
I went with `blockly-test` to keep it visually distinct from how the Blockly package is normally imported. Also, I figured it'd be easier to find `blockly-test` so those looking at this later could easily tell that the dist files are what is being tested.

cpcallen replied:
Those are both good reasons.  SGTM.

btw17 commented on tests/typescript/src/field/different_user_input.ts:
```diff
@@ -0,0 +1,75 @@
+    // Example-specific classes.
+
+    private isCell(maybeCell: unknown): maybeCell is Cell {
+        if (!maybeCell) return false;
+        const couldBeCell = maybeCell as { [key: string]: unknown };
+        return 'cellId' in couldBeCell && typeof couldBeCell.cellId === 'string';
```
Whoops! Perhaps my comments above mistakenly mentioning `classes` instead of `methods` may have contributed to this misunderstanding. 

Regarding using an `interface` vs a `class`, I don't think a class really makes sense here for either `Cell` or `CellGroup`.  Although there is the extra `isCell` function, we don't have to invoke a `new Cell` every time one needs to be created and we don't need to setup a constructor and methods for the `Cell`/`CellGroup` objects to create them.  Also, classes are ideal when you have complex state that needs to be maintained and/or when you have methods/values that are valuable for a subclass to inherit as they are. In this case, the purpose of the example is that `FieldMitosis` manages the state of these values. If we wanted to use a class, there should really be one class that encompasses `Cell` and `CellGroup` and handles methods like `doMitosis`, and that would invalidate this example.

Part of this may be personal preference, though, as I tend to avoid classes when possible. Classes are great in JavaScript when there's a bunch of uncertainty since everything is in one spot and you already have to keep everything involved in mind since you can't always trust that things are what they say they are. In TypeScript, though, using functions + many interfaces is great since you only have to keep in mind what the state is at that one point in time. If state should change, just invoke a function to transform the state to what it should be and have the resulting type reflect that. Makes functions super simple, too, since they just worry about the state coming in and how to produce the state going out.

btw17 commented on tests/typescript/src/field/different_user_input.ts:
```diff
@@ -0,0 +1,75 @@
+ * constructor input type than the type that is stored.
+ */
+import { Field, FieldValidator, fieldRegistry } from 'blockly-test/core';
+
```
Fixed!

cpcallen commented on tests/typescript/src/field/different_user_input.ts:
```diff
@@ -0,0 +1,75 @@
+
+/** Example use of the class. */
+
+const cellValidator: FieldMitosisValidator = (cellGroup) => {
```
> * Sounds good! Updated to just be a `function` instead of a `const`.

Did my suggestion actually type check successfully?  😃  Maybe I'm getting the hang of this TypeScript thing…

btw17 commented:
Should be all set! Please check my changes which update `.eslintrc.json` to `.eslintrc.js`. Other than restructuring the file for JS, I did the following updates:
- Added an ignore pattern for `.eslint.js` (which was being picked up otherwise and triggered an error due to using `module.exports`)
- Modified the file pattern for the override connected to `./tsconfig.json` to just refer to `./core` TS files instead of all TS files
- Added the override for `./tests/typescript/tsconfig.json`

cpcallen commented on scripts/gulpfiles/test_tasks.js:
```diff
@@ -371,6 +370,16 @@ function advancedCompileInBrowser() {
+ */
+function typeDefinitions() {
+  return runTestCommand('type_definitions',
+    'tsc -p ./tests/typescript/tsconfig.json');
```
Please write output to somewhere in `build/` rather than in `tests/`.

I think it would be preferable to do so using `-outDir` and `path.join(BUILD_DIR, 'tests' /* optionally more */)` as `buildJavaScript` (in `build_tasks.js`) does.

cpcallen commented on .eslintignore:
```diff
@@ -11,6 +11,7 @@
 /tests/screenshot/*
 /tests/test_runner.js
 /tests/workspace_svg/*
+/tests/typescript/dist/*
```
This should not be necessary.

btw17 replied:
When this is not ignored, I get an error like below:
```
5:1  error  Parsing error: 'import' and 'export' may appear only with 'sourceType: module'
```

Basically it's trying to run lint on the JS files that are created in `/tests/typescript/dist`

cpcallen commented on .eslintrc.js:
```diff
@@ -0,0 +1,197 @@
+}
+
+module.exports = {
+    "ignorePatterns": [".eslintrc.js"],
```
This should not be necessary.

btw17 replied:
I was able to fix this via providing an override for `./.eslintrc.js` that enabled building for a node environment.

cpcallen reviewed (changes requested):
In addition to changing the `tsc` output directory (see below), can you update `.eslintrc.js` to follow the JS style guide, please?  E.g.:

* Indentation by two spaces.
* Use single quotes for strings, no quoting of property names that are valid identifieres,
* Ends with `\n`.

You can probably just run `clang-format` on it.  Better: modify the config so that this file is always formatted when `npm run format` is run.

cpcallen commented on .eslintrc.js:
```diff
@@ -0,0 +1,215 @@
+    'goog' : true,
+    'exports' : true,
+  },
+  'extends' :
+            [
+              'eslint:recommended',
+              'google',
+            ],
```
Go home formatter, you're drunk:

```suggestion
  extends: [
    'eslint:recommended',
    'google',
  ],
```
(and similarly for `overrides`.)

cpcallen commented on .eslintrc.js:
```diff
@@ -0,0 +1,215 @@
+const rules = {
+  'curly': ['error'],
```
Single quotes not required for property names that are valid identifiers:
```suggestion
  curly: ['error'],
```
(and many below).

btw17 replied:
done

btw17 replied:
Looks like the linter wants consistently quoted properties. In TypeScript, it's best practice to quote keys you don't know at compile time. For example:

```ts
interface Test {
  knownKey: string;
  [key: string]: number;
}
const test: Test = {
  knownKey: 'value',
  'unknownOne': 1,
  'unknownTwo': 2
};
```

With that principle in mind, maybe it's valid to keep the keys that are defined by eslint outside of this file quoted.

cpcallen commented on scripts/gulpfiles/build_tasks.js:
```diff
@@ -733,7 +733,8 @@ function cleanBuildDir() {
     'core/**/*.js', 'core/**/*.ts',
-    'blocks/**/*.js', 'blocks/**/*.ts'
+    'blocks/**/*.js', 'blocks/**/*.ts',
+    '.eslintrc.js'
```
If you can't get clang-format to behave itself feel free to remove this and just fix the formatting manually.

btw17 replied:
Got it to work via moving the JSON logic to a `const` away from `module.exports`

cpcallen commented on scripts/gulpfiles/config.js:
```diff
@@ -38,6 +38,10 @@ exports.TYPINGS_BUILD_DIR = path.join(exports.BUILD_DIR, 'declarations');
 // Matches the value in tsconfig.json: outDir
 exports.TSC_OUTPUT_DIR = path.join(exports.BUILD_DIR, 'src');
 
+// Directory where typescript compiler can be found for `./tests/typescript`
+// files.
```
This description seems confusing to me.  How about:
```suggestion
// Directory for files generated by compiling test code.
```

btw17 replied:
done
